lib/common/xml.c

```diff
@@ -2056,6 +2027,92 @@ pcmk__xe_find_replace(xmlNode *xml, xmlNode *replace)
     return ENXIO;
 }
 
+/*!
+ * \internal
+ * \brief Update one XML subtree with another if the two match
+ *
+ * "Update" means to make a target subtree match a source subtree in children
+ * and attributes, recursively. \c "++" and \c "+=" in attribute values are
+ * expanded where appropriate (see \c expand_plus_plus()).
+ *
+ * A match is defined as follows:
+ * * \p xml and \p user_data are both element nodes of the same type.
+ * * \p xml and \p user_data have the same \c PCMK_XA_ID attribute value, or
+ *   \c PCMK_XA_ID is unset in both
+ *
+ * \param[in,out] xml        XML subtree to update with \p user_data upon match
+ * \param[in]     user_data  XML to update \p xml with upon match
+ *
+ * \return \c true to continue traversing the tree, or \c false to stop (because
+ *         \p xml was updated by \p user_data)
+ *
+ * \note This is compatible with \c pcmk__xml_foreach_dfs().
+ */
+static bool
+update_matching_xe(xmlNode *xml, void *user_data)
+{
+    xmlNode *update = user_data;
+
+    if (!pcmk__xe_is(update, (const char *) xml->name)) {
+        // No match: either not both elements, or different element types
+        return true;
+    }
+
+    if (!pcmk__str_eq(pcmk__xe_id(xml), pcmk__xe_id(update), pcmk__str_none)) {
+        // No match: ID mismatch
+        return true;
+    }
+
+#if XML_PARSER_DEBUG
+    crm_log_xml_trace(xml, "update-match");
+    crm_log_xml_trace(replace, "update-with");
+#endif  // XML_PARSER_DEBUG
+    pcmk__xml_update(NULL, xml, update, false);
+
+    // Found a match and replaced it; stop traversing tree
+    return false;
+}
+
+/*!
+ * \internal
+ * \brief Search an XML tree depth-first and update the first matching element
+ *
+ * "Update" means to make a target subtree match a source subtree in children
+ * and attributes, recursively. \c "++" and \c "+=" in attribute values are
+ * expanded where appropriate (see \c expand_plus_plus()).
+ *
+ * A match with a node \c node is defined as follows:
+ * * \c node and \p update are both element nodes of the same type.
+ * * \c node and \p user_data have the same \c PCMK_XA_ID attribute value, or
+ *   \c PCMK_XA_ID is unset in both
+ *
+ * \param[in,out] xml     XML tree to search
+ * \param[in]     update  XML to update a matching element with
+ *
+ * \return Standard Pacemaker return code (specifically, \c pcmk_rc_ok on
+ *         successful update and an error code otherwise)
+ */
+int
+pcmk__xe_find_update(xmlNode *xml, xmlNode *update)
+{
+    /* @COMPAT In pcmk__xe_find_replace() and pcmk__xe_find_delete(), we compare
+     * IDs only if the equivalent of the update argument has an ID. Here, we're
+     * stricter: we consider it a mismatch if only one element has an ID
+     * attribute, or if both elements have IDs but they don't match.
+     *
+     * Perhaps we should align the behavior at a major version release.
```

Comment:
It should be reasonable to follow the behavior here in all of them. It's likely laziness in the others

Reply:
It's technically a behavioral change if we do, since these are exposed indirectly via the CIB API.

Reply:
yeah that needs careful consideration. Aside from the fact that it would be a behavioral change in the public API, we might have legitimate places where we find and delete/replace the first element of a given type (or matching some arbitrary subset of attributes, in the case of delete).